syntax = "proto2";

message Attribute {
  required int32 type = 1;
  required string name = 2;
  oneof values {
    string str = 3;
    int32 integer = 4;
    float real = 5;
    int32 int32ean = 6;
  }
}

message Operator {
  required int32 field = 1;
  required Attribute value = 2;
}

message Comparator {
  required int32 negative = 1;
  required int32 is_true = 2;
  required int32 operation = 3;
  required Operator op1 = 4;
  required Operator op2 = 5;
}

message Filter {
  required int32 negative = 1;
  repeated Comparator comp = 2;
}

message Level {
  required int32 negative = 1;
  required int32 any = 2;
  required int32 parent = 3;
  required int32 id = 4;
  repeated Filter filters = 5;
}

message Query_tree {
  required int32 command = 1;
  repeated Level level = 2;
  required string name = 3;
  required int64 value = 4;
}

message Return_code {
  required int32 code = 1;
}

message Entity {
  repeated Attribute values = 1;
}

message Tuple {
  required Entity data = 1;
  required int32 id = 2;
  required int32 parent_id = 3;
}

message Collection {
  repeated Tuple tuples = 1;
}